### PR TITLE
Charts refactor

### PR DESCRIPTION
**Pull request by Qizot**
## Refactor
Refactored way of preparing charts and their series. Now each chart can cache its accumulator that can be used to somehow properly update the chart in case of a real-time update (caching last higest value for aggragate chart, caching all values for the last second for changes per second chart ect.).

## Supported charts
* `simple`  - plain line chart displaying unprocessed values for each timeline timestamp
* `cumulative` - it is an aggregation line chart, displaying sum of all previous values for each timestamp (`<from, current_timestamp>`)
* `changes_per_second` - similar to cumulative but the sum is kept just for the last second for each timestamp

## New metrics
Added new charts for `buffer`, `queue_len` and `bitrate` metrics;

### CODE REVIEW

File: lib/membrane_dashboard/dagre/g6_marshaller.ex [hunk 1]
```diff
@@ -156,7 +156,7 @@ defmodule Membrane.Dashboard.Dagre.G6Marshaller do
     [label | parents] = path |> Enum.reverse()
 
     parent_id =
-      if length(parents) == 0 do
+      if parents == [] do
```
Comment (sheldak) [inline lib/membrane_dashboard/dagre/g6_marshaller.ex:159]: I think `case` suits better here

File: lib/membrane_dashboard_web/live/dashboard_live.ex [hunk 2]
```diff
@@ -67,14 +69,16 @@ defmodule Membrane.DashboardWeb.DashboardLive do
 
     # if we have no data assigned then request the full query
     if socket.assigns.data == empty_lists do
-      {:noreply, push_patch_with_params(socket, %{from: from, to: to})}
+      push_patch_with_params(socket, %{from: from, to: to}) |> noreply()
```
Comment (sheldak) [inline lib/membrane_dashboard_web/live/dashboard_live.ex:72]: Just one pipe operator is not recommended so I would rewrite it to 
```suggestion
      socket
      |> push_patch_with_params(%{from: from, to: to}) 
      |> noreply()
```

File: lib/membrane_dashboard_web/live/dashboard_live.ex [hunk 3]
```diff
@@ -67,14 +69,16 @@ defmodule Membrane.DashboardWeb.DashboardLive do
...
     else
       with true <- connected?(socket),
            {from, to} <- extract_time_range(%{"from" => from, "to" => to}, socket) do
-        {:noreply, launch_query_task(socket, socket.assigns, from, to, :update)}
+        launch_query_task(socket, socket.assigns, from, to, :update) |> noreply()
```
Comment (sheldak) [inline lib/membrane_dashboard_web/live/dashboard_live.ex:76]: Same here (pipe)

File: lib/membrane_dashboard_web/live/dashboard_live.ex [hunk 4]
```diff
@@ -86,152 +90,177 @@ defmodule Membrane.DashboardWeb.DashboardLive do
...
     end
   end
 
   # inits, realoads or updates charts and reloads dagre
   @impl true
   def handle_info({:charts_init, data}, socket) do
-    {:noreply, push_event(socket, "charts_init", %{data: data})}
+    socket
+    |> push_event("charts_init", %{data: data})
+    |> noreply()
```
Comment (sheldak) [inline lib/membrane_dashboard_web/live/dashboard_live.ex:118]: Good idea with this `noreply()`! I really like it :D

File: lib/membrane_dashboard_web/live/dashboard_live.ex [hunk 5]
```diff
@@ -86,152 +90,177 @@ defmodule Membrane.DashboardWeb.DashboardLive do
...
+    socket
+    |> assign(time_from: time_from, time_to: time_to)
+    |> case do
+      # if the timer had not been set then do nothing
+      %{assigns: %{update_ref: nil}} = socket ->
+        socket
 
-    {:noreply, socket}
+      socket ->
+        socket |> plan_update()
```
Comment (sheldak) [inline lib/membrane_dashboard_web/live/dashboard_live.ex:145]: No need of pipe
```suggestion
        plan_update(socket)
```

**Review by sheldak — CHANGES_REQUESTED** [review]

File: .credo.exs [hunk 6]
```diff
@@ -0,0 +1 @@
+
```
[the remark this answers is not in the document]
Comment (Qizot) [reply .credo.exs:1]: Interesting...

File: lib/membrane_dashboard/dagre/g6_marshaller.ex (hunk 1 above)
Comment (Qizot) [reply lib/membrane_dashboard/dagre/g6_marshaller.ex:159]: I think that in this case `if` is pretty much equivalent.

File: lib/membrane_dashboard/charts/helpers.ex [hunk 7]
```diff
@@ -49,68 +52,70 @@ defmodule Membrane.Dashboard.Charts.Helpers do
...
+    rows
+    |> rows_to_data_by_paths()
+    |> process_changes_per_second_series(interval, initial_accumulator)
+  end
 
-    data_by_paths_to_series(data_by_paths, interval, initial_accumulators)
+  @doc """
+  Creates an aggregate series with an increasing values (each value is a sum of all previous and current value).
+  """
+  def to_cumulative_series(rows, interval, initial_accumulator) do
```
[the remark this answers is not in the document]
Comment (Qizot) [reply lib/membrane_dashboard/charts/helpers.ex:116]: Yup, just forgot about it.

File: lib/membrane_dashboard/dagre/g6_marshaller.ex (hunk 1 above)
Comment (sheldak) [reply lib/membrane_dashboard/dagre/g6_marshaller.ex:159]: Maybe it is a matter of opinion. So it can stay if you want

**Review by sheldak — APPROVED** [review]
LGTM!

**Pull request merged**